### PR TITLE
MAINT: special: use reflection formulas for spherical Bessel functions with negative real `z`

### LINKED ISSUE #14582: Negative arguments for scipy.special.spherical_jn leads to nans
`scipy.special.spherical_jn` seems to generate `nan`s for negative input arguments, which does not seem to be prohibited and the [documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.spherical_jn.html) does not seem to warn against. For example, the following code

```
import numpy as np
from scipy.special import spherical_jn

n = 5
zs = np.linspace(-5, 5, 10)
print(spherical_jn(n, zs))
```
yields an output of 
`[           nan            nan            nan            nan      nan 5.03098996e-06 1.11092345e-03 1.17516973e-02 4.66525179e-02 1.06811161e-01]`

Based on the reflection formula from [DLMF](http://dlmf.nist.gov/10.47.E14), I used the following fix for my purposes

```
import numpy as np
from scipy.special import spherical_jn

def spherical_bessel(n, z):
  return np.where(not isinstance(z, complex) and z<0, (-1)**n*spherical_jn(n, -z), spherical_jn(n, z))

zs = np.linspace(-5, 5, 10)
print(spherical_bessel(n, zs))
```
which gave the result
`[-1.06811161e-01 -4.66525179e-02 -1.17516973e-02 -1.11092345e-03
 -5.03098996e-06  5.03098996e-06  1.11092345e-03  1.17516973e-02
  4.66525179e-02  1.06811161e-01]`

I suspect that a similar, but more general fix can be used for a PR if this is deemed a legitimate issue. 

(I'm using Python 3.7.7, numpy-1.21.1 and scipy-1.7.1)

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-14582
Supersedes gh-14590 (but preserves those commits)

#### What does this implement/fix?
gh-14582 reported:
> `scipy.special.spherical_jn` seems to generate nans for negative input arguments, which does not seem to be prohibited and the documentation does not seem to warn against.

gh-14590 implemented reflection formulas to fix this for `spherical_jn` and the other spherical Bessel functions.

This re-implements the reflection formulas for both the functions and their derivatives in Python (since the underlying Cython files were deleted) and simplifies the tests. I chose a non-invasive (decorator) approach so that the implementation is easy to remove when there is time to implement the reflection formulas in the underlying C++ files. I opened a new PR because I can't push to the original branch, but we'll make sure the original author gets commit credit.

### CODE REVIEW

**Comment by mdhaber** [comment]
The failing test checks for agreement between the Python and Cython versions of the function. I didn't know it was exposed via Cython, so I didn't fix that, and it still returns NaN. LMK what you want to do with that.

File: scipy/special/_spherical_bessel.py [hunk 1]
```diff
@@ -1,8 +1,41 @@
...
+        def wrapper(n, z, derivative=False):
+            z = np.asarray(z)
+
+            if np.issubdtype(z.dtype, np.complexfloating):
+                return fun(n, z, derivative)  # complex dtype just works
+
+            f2 = standard_reflection if reflection_fun is None else reflection_fun
+            return _lazywhere(z.real >= 0, (n, z),
+                              f=lambda n, z: fun(n, z, derivative),
+                              f2=lambda n, z: f2(n, z, derivative))
```
Comment (mdhaber) [inline scipy/special/_spherical_bessel.py:33]:
```suggestion
                              f2=lambda n, z: f2(n, z, derivative))[()]
```

**Comment by steppi** [comment]
I think it's fine to `xfail` the relevant `test_cython_special` tests for now. Just change `None`s here to strings giving the failure reason:

https://github.com/scipy/scipy/blob/8ca4ace96cff522024b84e227b24b4324491faa1/scipy/special/tests/test_cython_special.py#L265-L272

I'll eventually follow up with a PR to move the reflection formulas down into the C++ code so the fix will work in `cython_special` too.

**Review by steppi — APPROVED** [review]
Thanks @mdhaber. Since we already are using Python wrappers here, I'm happy with getting a fix in at the Python level for the short term. I can follow up with a PR that moves the reflection formulas into the C++ code, so we can get this working in `cython_special` too.

The tests looks good. I'll merge after the linter's complaint is addressed.

File: scipy/special/tests/test_mpmath.py [hunk 2]
```diff
@@ -2239,8 +2239,9 @@ def mp_spherical_in(n, z):
 
     def test_spherical_kn(self):
         def mp_spherical_kn(n, z):
-            out = (mpmath.besselk(n + mpmath.mpf(1)/2, z) *
-                   mpmath.sqrt(mpmath.pi/(2*mpmath.mpmathify(z))))
+            arg = mpmath.mpmathify(z)
+            out = (mpmath.besselk(n + mpmath.mpf(1)/2, arg) /
+                   mpmath.sqrt(2*arg/mpmath.pi))
```
Comment (mdhaber) [inline scipy/special/tests/test_mpmath.py:2244]: One of those dark corners of floating point math that I hadn't encountered before:
```python3
import math
from mpmath import mp
x = -1 + 0j
((1/x)**0.5).imag  # -1.0
((1/x+0)**0.5).imag  # 1.0
((1/x)**mp.mpf(0.5)).imag  # mpf('1.0')
```

Apparently our complex implementation uses the negative (imaginary) square root.

**Comment by mdhaber** [comment]
> The tests looks good. I'll merge after the linter's complaint is addressed.

Done. Thanks for the review @steppi.

**Comment by mdhaber** [comment]
The CircleCI failure doesn't look real. TBH it wasn't supposed to run because I used `[lint only]`.

**Comment by steppi** [comment]
Thanks @mdhaber. Yeah, failures must be unrelated.

File: scipy/special/tests/test_mpmath.py (hunk 2 above)
Comment (steppi) [reply scipy/special/tests/test_mpmath.py:2244]: Cool, right? Check out http://port70.net/~nsz/articles/float/kahan_branch_cuts_complex_elementary_functions_1987.pdf if you’re interested

**Comment by WarrenWeckesser** [comment]
I get a new XSLOW failure in `scipy/special/tests/test_mpmath.py::TestSystematic::test_spherical_jn` after this change:

```
$ SCIPY_XSLOW=1 python3 dev.py test -m full -t scipy/special/tests/test_mpmath.py::TestSystematic::test_spherical_jn
💻  ninja -C /home/warren/repos/git/forks/scipy/build -j12
ninja: Entering directory `/home/warren/repos/git/forks/scipy/build'
[1/1] Generating scipy/generate-version with a custom command
Build OK
💻  meson install -C build --only-changed --tags runtime,python-runtime,tests,devel
Installing, see meson-install.log...
Installation OK
SciPy from development installed path at: /home/warren/repos/git/forks/scipy/build-install/lib/python3.12/site-packages
Running tests for scipy version:1.15.0.dev0+git20240926.4c936c8, installed at:/home/warren/repos/git/forks/scipy/build-install/lib/python3.12/site-packages/scipy
========================================= test session starts =========================================
platform linux -- Python 3.12.4, pytest-8.2.2, pluggy-1.5.0
rootdir: /home/warren/repos/git/forks/scipy
configfile: pytest.ini
plugins: hypothesis-6.105.1, timeout-2.3.1, xdist-3.6.1, cov-5.0.0, anyio-4.4.0
collected 1 item                                                                                      

scipy/special/tests/test_mpmath.py F                                                            [100%]

============================================== FAILURES ===============================================
__________________________________ TestSystematic.test_spherical_jn ___________________________________
scipy/special/tests/test_mpmath.py:2150: in test_spherical_jn
    assert_mpmath_equal(
        mp_spherical_jn = <function TestSystematic.test_spherical_jn.<locals>.mp_spherical_jn at 0x7dfe51fad3a0>
        self       = <scipy.special.tests.test_mpmath.TestSystematic object at 0x7dfe4f5184a0>
scipy/special/_mptestutils.py:295: in assert_mpmath_equal
    d.check()
        a          = (<function TestSystematic.test_spherical_jn.<locals>.<lambda> at 0x7dfe4f528e00>, <function exception_to_nan.<locals>....cipy.special._mptestutils.IntArg object at 0x7dfe51f0cb90>, <scipy.special._mptestutils.Arg object at 0x7dfe51ffbe30>])
        d          = <MpmathData: wrap>
        kw         = {'dps': 300}
scipy/special/_mptestutils.py:282: in check
    raise value
        argarr     = array([[ 0.00000000e+00, -1.00000000e+08],
       [ 0.00000000e+00, -1.33352143e+07],
       [ 0.00000000e+00, -1.7782....65000000e+02,  1.77827941e+06],
       [ 1.65000000e+02,  1.33352143e+07],
       [ 1.65000000e+02,  1.00000000e+08]])
        dps        = 300
        dps_list   = [300]
        j          = 0
        mptype     = <function MpmathData.check.<locals>.mptype at 0x7dfe4f528fe0>
        old_dps    = 15
        old_prec   = 53
        pytype     = <function MpmathData.check.<locals>.pytype at 0x7dfe4f529080>
        self       = <MpmathData: wrap>
        tb         = <traceback object at 0x7dfe4f540380>
        tp         = <class 'AssertionError'>
        value      = AssertionError('\nMax |adiff|: 2.28406e-294\nMax |rdiff|: 1\nBad results (2 out of 1820) for the following points (in ...78506e-21 =>                             0. !=         8.080520443892829e-298  (rdiff                             1.)')
scipy/special/_mptestutils.py:263: in check
    assert_func_equal(
        argarr     = array([[ 0.00000000e+00, -1.00000000e+08],
       [ 0.00000000e+00, -1.33352143e+07],
       [ 0.00000000e+00, -1.7782....65000000e+02,  1.77827941e+06],
       [ 1.65000000e+02,  1.33352143e+07],
       [ 1.65000000e+02,  1.00000000e+08]])
        dps        = 300
        dps_list   = [300]
        j          = 0
        mptype     = <function MpmathData.check.<locals>.mptype at 0x7dfe4f528fe0>
        old_dps    = 15
        old_prec   = 53
        pytype     = <function MpmathData.check.<locals>.pytype at 0x7dfe4f529080>
        self       = <MpmathData: wrap>
        tb         = <traceback object at 0x7dfe4f540380>
        tp         = <class 'AssertionError'>
        value      = AssertionError('\nMax |adiff|: 2.28406e-294\nMax |rdiff|: 1\nBad results (2 out of 1820) for the following points (in ...78506e-21 =>                             0. !=         8.080520443892829e-298  (rdiff                             1.)')
scipy/special/_testutils.py:85: in assert_func_equal
    fdata.check()
E   AssertionError: 
E   Max |adiff|: 2.28406e-294
E   Max |rdiff|: 1
E   Bad results (2 out of 1820) for the following points (in output 0):
E                              12.          -4.024587945074773e-24 =>                             0. !=        2.2840641502320017e-294  (rdiff                             1.)
E                              14.          -8.073877075678506e-21 =>                             0. !=         8.080520443892829e-298  (rdiff                             1.)
        atol       = 1e-300
        data       = array([[ 0.00000000e+00, -1.00000000e+08],
       [ 0.00000000e+00, -1.33352143e+07],
       [ 0.00000000e+00, -1.7782....65000000e+02,  1.77827941e+06],
       [ 1.65000000e+02,  1.33352143e+07],
       [ 1.65000000e+02,  1.00000000e+08]])
        distinguish_nan_and_inf = True
        dtype      = None
        fdata      = <Data for <lambda>>
        func       = <function TestSystematic.test_spherical_jn.<locals>.<lambda> at 0x7dfe4f528e00>
        ignore_inf_sign = False
        knownfailure = None
        nan_ok     = True
        nparams    = 2
        param_filter = None
        points     = array([[ 0.00000000e+00, -1.00000000e+08],
       [ 0.00000000e+00, -1.33352143e+07],
       [ 0.00000000e+00, -1.7782....65000000e+02,  1.77827941e+06],
       [ 1.65000000e+02,  1.33352143e+07],
       [ 1.65000000e+02,  1.00000000e+08]])
        result_columns = None
        result_func = <function MpmathData.check.<locals>.<lambda> at 0x7dfe4f529120>
        results    = <function MpmathData.check.<locals>.<lambda> at 0x7dfe4f529120>
        rtol       = 1e-07
        vectorized = False
======================================= short test summary info =======================================
FAILED scipy/special/tests/test_mpmath.py::TestSystematic::test_spherical_jn - AssertionError: 
========================================== 1 failed in 0.72s ==========================================

```

**Comment by mdhaber** [comment]
We'll just have to skip the test. This simply exposes some existing (slightly) premature underflow.

```python3
import numpy as np
from scipy import special
import matplotlib.pyplot as plt

x = np.logspace(-30, -1)
y = special.spherical_jn(12, x)
plt.loglog(x, abs(y))
```
{2FC1E571-F58E-4210-BC4E-3EAA79DF4B40}

**Pull request merged**